### DISCUSSION TITLE
dofs in subdomain

### DISCUSSION THREAD

**Discussion by gatling-nrl** [kinnala/scikit-fem#764] [Show and tell]
In my application (solving Laplace's equation for electric potential) I have some embedded electrodes. Some of them act as boundary conditions and some act as probes. Each electrode occupies its own subdomain in the mesh. I want to get the dofs in these subdomains to set the boundary conditions, or in the case of the probes, compute the average potential.

I've got a working solution, I just think it's ugly and maybe not that fast. Ideas for improvement? Even though facets use midpoints and so I used midpoints here too, I don't trust `==` with floats. Hence the `np.linalg.norm` is acting like a broadcastable `isclose`.

```python
def get_dofs_in_subdomain(basis, subdomain):
    eps = 1e-8
    fem_mesh = basis.mesh
    subdomain_midpts = list()
    for cell_idx in fem_mesh.subdomains[subdomain]:
        topo = fem_mesh.t.T[cell_idx]
        for edge in [[0,1],[1,2],[2,0]]: # connectivity of one topo element
            subdomain_midpts.append(fem_mesh.p.T[topo[edge]].sum(axis=0)/2.0)
    subdomain_midpts = np.array(subdomain_midpts)            
    facets = fem_mesh.facets_satisfying(
        lambda x: [(np.linalg.norm(subdomain_midpts-x0, axis=1) < eps).any()
            for x0 in x.T]
    )
    return basis.get_dofs(facets)
```

**Comment by gdmcbain** [comment]
How about ex26?

https://github.com/kinnala/scikit-fem/blob/7500d1ca3523196b8d88e6c2cbf7cfecd6c52331/docs/examples/ex26.py#L34

**Reply by gatling-nrl** [reply]
I think that's a lot prettier! Thanks! Is that common enough to belong in e.g. `basis.get_dofs(subdomain='annulus')`?

**Reply by gdmcbain** [reply]
That's an idea, yes.

Or even `basis['annulus']`, if we were to define the [`__getitem__`](https://docs.python.org/3/reference/datamodel.html?highlight=__getitem__#object.__getitem__) method.  (I haven't thought that through at all.)

I'm still not sure about `find_dofs` versus `get_dofs`, whether one will eventually be deprecated.  There are many more uses of the former in `docs/examples` but the later additions seem to prefer the latter.

**Reply by gatling-nrl** [reply]
@gdmcbain So I gave that a try, but they're not quite equivalent. `basis.get_dofs` returns a `DofsView` but `annulus` contains a `ndarray` after line 34. Later in my code `DofsView + ndarray` fails where `DofsView + DofsView` previously succeeded.

Digging around it seems like `DofsView` is (_almost_?) always treated like an array and that it can flattened as soon as its gotten and concatenated with other `ndarray`s in the usual ways. Am I missing something with that?

**Reply by gatling-nrl** [reply]
Personally, I would find the `basis['annulus']` notation confusing because I don't think of the basis as a collection of dofs. I'd intuitively expect basis['annulus'] to return an set of elements or something like that.  `basis['annulus'].get_dofs()` ?

**Reply by gatling-nrl** [reply]
Since `DofsView` behaves a lot like `ndarray` the following seems to work. 
```python
bottom_dofs = basis.get_dofs(fem_mesh.facets_satisfying(lambda x: np.isclose(x[1], 0))
sub0_dofs = np.unique(basis.element_dofs[:, fem_mesh.subdomains['subdomain0_name']])
sub1_dofs = np.unique(basis.element_dofs[:, fem_mesh.subdomains['subdomain1_name']])
boundary_dofs = np.hstack([bottom_dofs, sub0_dofs, sub1_dofs])
boundary_values = basis.zeros()
boundary_values[[sub1_dofs]] = 1.0 # e.g. volts
A,b = skfem.enforce(A, D=boundary_dofs, x=boundary_values)
u = skfem.solve(A, b)
```

**Comment by gatling-nrl** [comment]
This little bit of syntactic sugar made the app code much more readable

```python
class MyBasis(skfem.Basis):

    def get_dofs(self, *args, **kwargs):
        if 'subdomain' in kwargs:
            mesh = self.mesh
            subdomain = mesh.subdomains[kwargs['subdomain']]
            return np.unique(self.element_dofs[:, subdomain])
        elif 'facets_satisfying' in kwargs:
            facets = self.mesh.facets_satisfying(kwargs['facets_satisfying'])
            return super().get_dofs(facets).flatten()

        else:
            return super().get_dofs(*args, **kwargs).flatten()
```

as in

```python
basis = MyBasis(skfem.io.from_meshio(mesh), skfem.ElementTriP1())

b_dofs = np.hstack([
    basis.get_dofs(facets_satisfying=lambda x: np.isclose(x[0], 0)), # left
    basis.get_dofs(facets_satisfying=lambda x: np.isclose(x[0], 1)), # right
    basis.get_dofs(facets_satisfying=lambda x: np.isclose(x[1], 0)), # bottom
    basis.get_dofs(subdomain='grounded_electrode'),
    basis.get_dofs(subdomain='source_electrode'),
])

**Reply by kinnala** [reply]
That's cool. I'd use it as such but sometimes you need to do other operations than `flatten()`, e.g., `all(['u^1'])` when you have multiple DOFs per node in vectorial problems (elasticity, flow problems) and constrain only x-component. Maybe it'd be enough to drop `flatten()` and then implement `DofsView.__array__()`?

**Reply by kinnala** [reply]
https://numpy.org/devdocs/user/basics.dispatch.html

> If we operate on `arr` with a numpy function, numpy will again use the `__array__` interface to convert it to an array and then apply the function in the usual way.

**Reply by gatling-nrl** [reply]
The problem I had was that `DofsView` had implemented `__add__` so that you could use `DofsView + DofsView` (as if they were lists)

**Reply by gatling-nrl** [reply]
Looking through the `DofsView` class, I see its a little more complicated than that.

**Reply by kinnala** [reply]
I think that `hstack`ing you can already do in the current implementation, I just didn't remember it:
```python
In [1]: from skfem import *                                                     

In [2]: m = MeshTri()                                                           

In [3]: basis = Basis(m, ElementVector(ElementTriP1()))                         

In [4]: import numpy as np                                                      

In [6]: np.hstack([ 
   ...: basis.get_dofs(lambda x: np.isclose(x[0], 0)), 
   ...: basis.get_dofs(lambda x: np.isclose(x[0], 1)), 
   ...: ])                                                                      
Out[6]: array([0, 1, 4, 5, 2, 3, 6, 7])
```

**Reply by kinnala** [reply]
> The problem I had was that DofsView had implemented __add__ so that you could use DofsView + DofsView (as if they were lists)

That's fair critique. Maybe `DofsView.__add__` is not that useful after all.

(Sorry for modifying your comment, I'm a bit tired.)

**Reply by gatling-nrl** [reply]
> That's cool. I'd use it as such but sometimes you need to do other operations than flatten(), e.g., all(['u^1']) when you have multiple DOFs per node in vectorial problems (elasticity, flow problems) and constrain only x-component. Maybe it'd be enough to drop flatten() and then implement DofsView.__array__()?

This seems to be okay, but I don't favor functions that have different return values. How to make the subdomain case return the same thing as `super().get_dofs` in the other cases?

It has turned out to be very convenient and precise to mark the subdomains when I'm creating the meshes instead of using `facets_satisfying`. Would you consider adding the subdomain feature to `get_dofs`?

**Reply by gdmcbain** [reply]
Yes, I usually define the subdomains before meshing, when the semantics are still in hand, rather than after too. Then I use `find_dofs` as in

https://github.com/kinnala/scikit-fem/blob/62882d61951e5283196cf22b9b9924b86f5b243b/docs/examples/ex24.py#L42

But as mentioned earlier i'm not sure that having two methods called `get_dofs` and `find_dofs` is ultimately what we'd want in the API. Not that I have a proposal to streamline that just yet.

**Reply by kinnala** [reply]
Sure, I'll do it. Not exactly as here since I think get_dofs should return DofsView. But the idea would be the same. Moreover, I think get_dofs(boundary='name') should also work, right?

**Reply by gdmcbain** [reply]
Is the idea to eventually deprecate `find_dofs`?

**Reply by gatling-nrl** [reply]
Oh, I'm not sure I realized `find_dofs()[subdomain_name].all()` worked like that. I have found `get_dofs` and `find_dofs` confusing, and when working from memory used the wrong one.

**Reply by gatling-nrl** [reply]
And yes, I agree should always return `DofsView`. I couldn't figure out how to get the `DofsView` after I had the result from `np.unique` which is why I used `flatten()` everywhere.

I'm not sure what `get_dofs(boundary='name')` gets me. What is 'name'? Is that something gmsh adds when it creates meshes? Maybe I just didn't realize it was possible to name mesh boundaries.

**Reply by kinnala** [reply]
> Is the idea to eventually deprecate `find_dofs`?

Right now it seems to be on the track of getting deprecated. Soon `get_dofs` will be able to do anything that `find_dofs` can.

**Reply by gatling-nrl** [reply]
```python
u = basis.zeros()
dofs = basis.get_dofs()
u[dofs.flatten()] = whatever
```
Are we trying to make `DofsView` act like `ndarray` when possible? Can we get rid of the need for `.flatten()` in this use case?

```python
u[dofs] = whatever
```
`IndexError: only integers, slices (':'), ellipsis ('...'), numpy.newaxis ('None') and integer or boolean arrays are valid indices`

It looks like if you define `__array_struct__`, `__array_interface__`, and `__array__` on `DofsView` you might be able to get it to pretend to be `ndarray` in most use cases. I can try to look into that if you think that's desirable behavior. I'm too new to skfem to anticipate what kinds of unintended side effects might happen or whether this would suppress user code exceptions and make actual bugs hard to find.

What was the rational for being a `namedTuple` instead of `ndarray`?

**Reply by kinnala** [reply]
`Dofs` object includes information about various things such as which topological entities the degrees-of-freedom are connected to (nodes, edges, facets, elements) and the global degree-of-freedom numbering as required in the finite element assembly.

`DofsView` is a "subset of DOFs" contained in a full `Dofs` object. It is not `ndarray`, it is a reference to a `Dofs` object, 4 index arrays and additional information about which `Element.dofnames` are included.

**Reply by kinnala** [reply]
> Can we get rid of the need for `.flatten()` in this use case?
> 
> ```python
> u[dofs] = whatever
> ```

Yeah it makes sense.

**Reply by gatling-nrl** [reply]
should it work for `Dofs` as well?

```python
u = basis.zeros()
type(a) # skfem.Dofs
u[a] = 1
type(b) # skfem.DofsView
u[b] = 2
```

What about iterators, etc?

```python
len(a) # <-- number of dofs in a?
len(b) # <-- number of dofs in b?
for dof in a:  # <-- would dof here be a integer? seems unintuitive though
    u[dof] = some_val
for dof in b: # <-- would dof here be a integer? seems unintuitive though
    ...
```

and does making `ndarray`s shaped like dofs make sense?

```python
np.zeros_like(a)
np.empty_like(b, dtype=...)
```

Do you think about `Dofs` and `DofsView` objects as a "list of dofs" like I'm doing here?

**Reply by gatling-nrl** [reply]
Would `basis.zeros()` become equivalent to `np.zeros_like(basis.get_dofs(), dtype=float)`?

edit: add `dtype=float`

**Reply by gdmcbain** [reply]
`basis.zeros()` would be `float` rather than `int` in terms of `dtype`.

**Reply by kinnala** [reply]
> should it work for `Dofs` as well?

No. Edit: `Dofs` is essentially all DOFs. So setting `u[all_dofs] = 1` is the same as `u[:] = 1`.

**Reply by kinnala** [reply]
Btw I added this in #775:
```python
from skfem import MeshTri, Basis, ElementTriP1
m = (MeshTri()
      .refined()
      .with_subdomains({'left': lambda x: x[0] < .5}))
basis = Basis(m, ElementTriP1())
basis.get_dofs(elements='left')  # 'elements' instead of subdomain
# to be consistent with the existing kwarg 'facets'
```
See the docstring for more info.

**Reply by gdmcbain** [reply]
O. K., good, that looks good.  Last time i looked `docs/examples` contained many more uses of `find_dofs` than `get_dofs` but with that `elements` keyword argument it would be easy to update and standardize. Thanks!

**Reply by gatling-nrl** [reply]
I'd like to suggest a couple of small changes....

```python
...
      if facets is not None and elements is not None:
          raise Exception('only one of `elements` or `facets` should'
                          'be specified')
    if elements is not None:
        ...
        elif isinstance(elements, (list, tuple)):
            elements = np.hstack([self.get_dofs(elements=e) for e in elements])
        ...
```

**Reply by kinnala** [reply]
I don't understand the latter. Why would you want to put DOF indices to `elements`? Do you mean something like:
```python
elif isinstance(elements, (list, tuple)):
    elements = np.unique(np.array([self.mesh.subdomains[e] for e in elements]).flatten())
```
in order to support
```python
basis.get_dofs(elements=["left", "right"])
```

**Reply by kinnala** [reply]
Support for multiple names is in #778.

**Reply by gatling-nrl** [reply]
>in order to support
>`basis.get_dofs(elements=["left", "right"])`

yes, that's exactly what i was thinking, except I was trying to be more general and also support
```python
basis.get_dofs(elements=["left", lambda x: x[0]>.5])
```

**Reply by kinnala** [reply]
That should work after #778, didn't test it though.
